### ISSUE TITLE
DALI 2021 roadmap

### ISSUE THREAD

**Issue by JanuszL** [NVIDIA/DALI#2978]
The following represents a high-level overview of our 2021 plan. You should be aware that this roadmap may change at any time and the order below does not reflect any type of priority.

We strongly encourage you to comment on our roadmap and provide us feedback on this issue here.

Improving Usability:
----------------------

- **[Done] improving functional API** - improving user experience by simplification of DALI functional API, grouping operations into modules, and extending examples to show how to use it - https://github.com/NVIDIA/DALI/pull/2761, https://github.com/NVIDIA/DALI/pull/2748, https://github.com/NVIDIA/DALI/pull/2745, https://github.com/NVIDIA/DALI/pull/2744, https://github.com/NVIDIA/DALI/pull/2743, https://github.com/NVIDIA/DALI/pull/2729, https://github.com/NVIDIA/DALI/pull/2725, https://github.com/NVIDIA/DALI/pull/2722, https://github.com/NVIDIA/DALI/pull/2721, https://github.com/NVIDIA/DALI/pull/2716, https://github.com/NVIDIA/DALI/pull/2713, https://github.com/NVIDIA/DALI/pull/2710, https://github.com/NVIDIA/DALI/pull/2653, https://github.com/NVIDIA/DALI/pull/2629, https://github.com/NVIDIA/DALI/pull/2566, https://github.com/NVIDIA/DALI/pull/2568
- **[In progress] eager mode** - introducing DALI operators callable as standalone entities to simplify debugging and prototyping - https://github.com/NVIDIA/DALI/pull/3531
- **[In progress] conditional execution** - ability to conditionally apply each operation, providing auto Augmentor style capabilities - https://github.com/NVIDIA/DALI/pull/3452, https://github.com/NVIDIA/DALI/pull/3451, https://github.com/NVIDIA/DALI/pull/3446, https://github.com/NVIDIA/DALI/pull/3403, https://github.com/NVIDIA/DALI/pull/3351, https://github.com/NVIDIA/DALI/pull/3339, https://github.com/NVIDIA/DALI/pull/3319, https://github.com/NVIDIA/DALI/pull/3311, https://github.com/NVIDIA/DALI/pull/3283, https://github.com/NVIDIA/DALI/pull/3280, https://github.com/NVIDIA/DALI/pull/3281, https://github.com/NVIDIA/DALI/pull/3276, https://github.com/NVIDIA/DALI/pull/3576, https://github.com/NVIDIA/DALI/pull/3701
- **intra-pipeline batch size variability** - providing an ability to change batch size from operator to operator inside the execution graph
  - [Done] Iteration-to-iteration batch size variability with non-parallel External Source - https://github.com/NVIDIA/DALI/pull/2641
- **[Done] tensor indexing** - https://github.com/NVIDIA/DALI/pull/3195

Extending input format support:
-----------------------------------

- enabling support of **variable frame rate videos**
  - [IN PROGRESS] https://github.com/NVIDIA/DALI/pull/3362, https://github.com/NVIDIA/DALI/pull/3391
- enabled **additional video formats**:
  - [DONE] VP8 and MJPEG - https://github.com/NVIDIA/DALI/pull/3045
- **parallelization of external source operators** to scale to multiple CPU cores
  - [Done] Parallel External Source for per-sample callbacks - https://github.com/NVIDIA/DALI/pull/2543
  - [Done] Support for lambdas and nested functions in Parallel External Source - https://github.com/NVIDIA/DALI/pull/3269
- **external source** support for TensorFlow - https://github.com/NVIDIA/DALI/pull/2997, https://github.com/NVIDIA/DALI/pull/3058, https://github.com/NVIDIA/DALI/pull/3089

Performance:
--------------

- **reducing memory consumption and allocation overhead** by introducing custom, pooled memory allocator - https://github.com/NVIDIA/DALI/pull/2930, https://github.com/NVIDIA/DALI/pull/2890, https://github.com/NVIDIA/DALI/pull/2891, https://github.com/NVIDIA/DALI/pull/2858, https://github.com/NVIDIA/DALI/pull/2814, https://github.com/NVIDIA/DALI/pull/2742, https://github.com/NVIDIA/DALI/pull/2609, https://github.com/NVIDIA/DALI/pull/2948, https://github.com/NVIDIA/DALI/pull/3536
- **acceleration for audio decoding** - providing acceleration for decoding various audio formats

New transformations:
------------------------

We are constantly extending the set of operations supported by DALI. Currently, this section lists the most notable additions to our areas of interest that were done this year. This list is not exhaustive and we plan on expanding the set of operators as the needs or requests arise. 

- new transformations for **images processing**
  - JpegCompressionDistortion operator - https://github.com/NVIDIA/DALI/pull/2823
  - MultiPaste operator - https://github.com/NVIDIA/DALI/pull/2681, https://github.com/NVIDIA/DALI/pull/2583
  - Gridmask operator - https://github.com/NVIDIA/DALI/pull/2652, https://github.com/NVIDIA/DALI/pull/2582
  - RandomObjectBBox operator - https://github.com/NVIDIA/DALI/pull/2657
  - ROIRandomCrop operator - https://github.com/NVIDIA/DALI/pull/2638
- new transformations for **audio processing**
  - time-major layout in the Spectrogram operator - https://github.com/NVIDIA/DALI/pull/2619, https://github.com/NVIDIA/DALI/pull/2617
  - time-major layout in the MelFilterBank GPU operator - https://github.com/NVIDIA/DALI/pull/2620
- new transformations for **3D/volumetric data processing**
  - RandomObjectBBox operator - https://github.com/NVIDIA/DALI/pull/2657
  - ROIRandomCrop operator - https://github.com/NVIDIA/DALI/pull/2638
- new transformations for **video processing**
- new **generic operations**
  - SaltAndPepper operator - https://github.com/NVIDIA/DALI/pull/2956, https://github.com/NVIDIA/DALI/pull/2889
  - ShotNoise operator - https://github.com/NVIDIA/DALI/pull/2861
  - more mathematical operations - https://github.com/NVIDIA/DALI/pull/2853
  - squeeze operator - https://github.com/NVIDIA/DALI/pull/2792
  - NumbaFunc operator - https://github.com/NVIDIA/DALI/pull/2804

**Comment by zhimengf** [comment]
I'm really looking forward to the support of YOLO v3/v4. Including,
1. multiscale-training. Input images' height and width are randomly generated and changes every N training steps. Currently DALI can only resize image to fixed size.  So this is not supported out-of-the-box by DALI
2. YOLO augmentations, hsv augmentation(YOLO style) and jitter(YOLO style, not color jitter)
3. YOLO label encoder

**Comment by JanuszL** [comment]
> I'm really looking forward to the support of YOLO v3/v4. Including,
> 
> 1. multiscale-training. Input images' height and width are randomly generated and changes every N training steps. Currently DALI can only resize image to fixed size.  So this is not supported out-of-the-box by DALI
> 2. YOLO augmentations, hsv augmentation(YOLO style) and jitter(YOLO style, not color jitter)
> 3. YOLO label encoder

1. In this case you can use the external source operator to load the parameters and pass them to the other operators. Mentioned resize operator accepts arguments for other operators - including mentioned external source.
2. Could you be more specific about what does it mean YOLO style - what DALI does and what you would expect to have?
3. We don't plan to add such an encoder as it is specific for YOLO. We have an encoder for SSD as this was the very first detection network DALI has started support, we would prefer to rely on the community to add encoders for other networks. Moreover, some implementations make the encoding the part of the network itself, not the data processing part.

**Comment by zhimengf** [comment]
Hi @JanuszL 
1. Could you give an short example on how to achieving this with DALI API? 
2. I just mean the hsv augmentation in yolo v3 is not the same as the hsv augmentation that is already implemented in DALI. So I'm expecting DALI to provide the exact hsv augmentation as in yolo v3.
3. That makes sense to me. Thanks!

**Comment by JanuszL** [comment]
Hi @zhimengf,
1) This simple example shows how you can use an external source to feed parameters to resize. You can read more about [the `external_source` usage in this example](https://docs.nvidia.com/deeplearning/dali/user-guide/docs/examples/general/data_loading/external_input.html).
```
import os
import numpy as np
import nvidia.dali.fn as fn
import nvidia.dali.types as types
from nvidia.dali import pipeline_def

batch_size = 64

def get_data():
    out = [(np.random.ranf(size=[2]).astype(dtype=np.float32)*60 + 30) for _ in range(batch_size)]
    return out

@pipeline_def
def simple_pipeline():
    jpegs, _ = fn.readers.file(file_root=os.environ['DALI_EXTRA_PATH'] + "/db/single/jpeg")
    images = fn.decoders.image(jpegs)
    size = fn.external_source(source=get_data)
    images = fn.resize(images, size=size)
    return images

pipe = simple_pipeline(batch_size=batch_size, num_threads=4, prefetch_queue_depth=2, device_id=0)
pipe.build()
out = pipe.run()[0]
print(np.array(out[0]).shape)
out = pipe.run()[0]
print(np.array(out[0]).shape)
out = pipe.run()[0]
print(np.array(out[0]).shape)
```
2) I'm sorry but I still don't understand where is the difference based on your description (I'm not familiar with how hsv/jitter works/is defined in YOLO v3). Could you describe the behavior you expect? Maybe it is just a different API and you can still use DALI but you need to express the transformation differently?

**Comment by JanuszL** [comment]
We have 2022 now. Closing this in favor of https://github.com/NVIDIA/DALI/issues/3774.

**Comment by carmocca** [comment]
The README roadmap section still points to this outdated issue: https://github.com/NVIDIA/DALI#dali-roadmap

**Comment by JanuszL** [comment]
@carmocca - good catch. https://github.com/NVIDIA/DALI/pull/3918 should fix that.
